### PR TITLE
Fix inefficient ActionListener.delegateFailure and ActionListener.wrap usage

### PR DESCRIPTION
**Pull request by original-brownbear**
ActionListener.delegateFailure is not the optimal choice when the response handler always invokes delegate.onResponse, it's just needlessly verbose, error-prone and sometimes less efficient when multiple map/delegation calls are chained. ActionListener.wrap is just a less efficient version of ActionListener.delegateAndWrap when the error handler is `listener::onFailure` -> fixup a couple of these as well.

### CODE REVIEW

**Review by DaveCTurner — COMMENTED** [review]
This is only a behaviour-preserving refactoring if none of the `BiConsumer` instances passed to `delegateFailure` ever throw. Bubbling the exception out to the `onResponse` caller is almost certainly a bug, but I'm worried it's a bug on which we rely somewhere. Could we introduce a variant of `map` which at least asserts the mapping function doesn't throw?

**Comment by original-brownbear** [comment]
> Could we introduce a variant of map which at least asserts the mapping function doesn't throw?

Wouldn't it make more sense to assert that we don't throw in the `delegateFailure` biconsumer? In fact ... when I opened this one I thought that was the current behavior but turns out that's only the case for `delegateResponse` which is a confusing asymetry isn't it?

**Comment by DaveCTurner** [comment]
Yeah we should assert that too.

File: server/src/main/java/org/elasticsearch/action/ActionListener.java [hunk 1]
```diff
@@ -69,6 +70,13 @@ default <T> ActionListener<T> map(CheckedFunction<T, Response, Exception> fn) {
         return new ActionListenerImplementations.MappedActionListener<>(fn, this);
     }
 
+    /**
+     * Same as {@link #map(CheckedFunction)} except that {@code fn} is expected to never throw.
+     */
+    default <T> ActionListener<T> safeMap(Function<T, Response> fn) {
+        return new ActionListenerImplementations.SafeMappedActionListener<>(fn, this);
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/action/ActionListener.java:77]: Should we fall back to `map()` if `Assertions.ENABLED == false`? Not sure if that extra try/catch is worth avoiding in prod or not.

**Comment by original-brownbear** [comment]
Jenkins run elasticsearch-ci/bwc

**Review by DaveCTurner — APPROVED** [review]
LGTM

**Pull request merged**